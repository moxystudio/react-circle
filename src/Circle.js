import React, { useRef, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const DIRECTIONS = {
    CLOCKWISE: 'clockwise',
    ANTI_CLOCKWISE: 'antiClockwise',
    BOTH_SIDES: 'bothSides',
};

const getSvgClassName = (strokePercentage, direction, className) => classNames(
    'react-circle_svg',
    `react-circle_${direction}`,
    {
        'react-circle_noStroke': strokePercentage === 0,
    },
    className,
);

const Circle = ({ className, direction, strokeWidth, strokePercentage, onTransitionEnd }) => {
    const svgRef = useRef(null);
    const [circleState, setCircleState] = useState();
    const [svgState, setSvgState] = useState({ className: getSvgClassName(strokePercentage, direction, className) });

    useEffect(() => {
        const { clientWidth: svgSize } = svgRef.current;

        const center = svgSize / 2;
        const radius = Math.max(center - parseInt(strokeWidth, 10), 0);
        const perimeter = 2 * Math.PI * radius;
        const rotate = (-strokePercentage * 180);
        const strokeLineSize = perimeter * strokePercentage;
        const strokeGapSize = perimeter - strokeLineSize;

        const strokeDasharray = `${strokeLineSize} ${strokeGapSize}`;

        const circleStyle = {
            strokeWidth,
            strokeDasharray,
        };
        const groupStyle = {};

        if (direction === DIRECTIONS.BOTH_SIDES) {
            groupStyle.transform = `rotate(${rotate}deg)`;
        }

        setCircleState({
            center,
            radius,
            perimeter,
            rotate,
            strokeGapSize,
            strokeLineSize,
            strokeDasharray,
            circleStyle,
            groupStyle,
        });

        setSvgState({ className: getSvgClassName(strokePercentage, direction, className) });
    }, [svgRef, strokeWidth, strokePercentage, direction, className]);

    return (
        <svg
            ref={ svgRef }
            className={ svgState.className }>
            { svgRef.current &&
                <g style={ circleState.groupStyle }>
                    <circle
                        style={ circleState.circleStyle }
                        cx={ circleState.center }
                        cy={ circleState.center }
                        r={ circleState.radius }
                        onTransitionEnd={ onTransitionEnd } />
                </g>
            }
        </svg>
    );
};

Circle.propTypes = {
    className: PropTypes.string,
    direction: PropTypes.oneOf(Object.values(DIRECTIONS)),
    strokeWidth: PropTypes.number,
    strokePercentage: PropTypes.number,
    onTransitionEnd: PropTypes.func,
};

Circle.defaultProps = {
    direction: DIRECTIONS.CLOCKWISE,
    strokeWidth: 1.5,
    strokePercentage: 1,
};

export default Circle;
