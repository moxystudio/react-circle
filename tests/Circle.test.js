import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Circle from '../src/Circle';

const renderWithProps = (props = {}, options) => render(<Circle { ...props } />, options);

describe('Circle Component', () => {
    it('should render correctly', () => {
        const { asFragment } = renderWithProps();

        expect(asFragment()).toMatchSnapshot();
    });

    it('should render correctly with direction both sides', () => {
        const { asFragment } = renderWithProps({ direction: 'bothSides' });

        expect(asFragment()).toMatchSnapshot();
    });

    it('should call onTransitionEnd callback prop', () => {
        const handleTransitionEnd = jest.fn();
        const { container } = renderWithProps({ onTransitionEnd: handleTransitionEnd });

        const circle = container.querySelector('circle');

        fireEvent.transitionEnd(circle);

        expect(handleTransitionEnd).toHaveBeenCalledTimes(1);
    });
});
