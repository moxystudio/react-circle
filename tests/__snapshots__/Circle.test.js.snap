// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle Component should render correctly 1`] = `
<DocumentFragment>
  <svg
    class="react-circle_svg react-circle_clockwise"
  >
    <g>
      <circle
        cx="0"
        cy="0"
        r="0"
        style="stroke-width: 1.5; stroke-dasharray: 0 0;"
      />
    </g>
  </svg>
</DocumentFragment>
`;

exports[`Circle Component should render correctly with direction both sides 1`] = `
<DocumentFragment>
  <svg
    class="react-circle_svg react-circle_bothSides"
  >
    <g
      style="transform: rotate(-180deg);"
    >
      <circle
        cx="0"
        cy="0"
        r="0"
        style="stroke-width: 1.5; stroke-dasharray: 0 0;"
      />
    </g>
  </svg>
</DocumentFragment>
`;
